src/librustc_trans/mir/mod.rs

```diff
@@ -123,26 +123,25 @@ impl<'blk, 'tcx> MirContext<'blk, 'tcx> {
         // locations of macro expansions with that of the outermost expansion site
         // (unless the crate is being compiled with `-Z debug-macros`).
         if source_info.span.expn_id == NO_EXPANSION ||
-           self.fcx.ccx.sess().opts.debugging_opts.debug_macros {
-            let scope_metadata = self.scope_metadata_for_span(source_info.scope, source_info.span);
+            source_info.span.expn_id == COMMAND_LINE_EXPN ||
+            self.fcx.ccx.sess().opts.debugging_opts.debug_macros {
+
+            let scope_metadata = self.scope_metadata_for_loc(source_info.scope,
+                                                             source_info.span.lo);
             DebugLoc::ScopeAt(scope_metadata, source_info.span)
         } else {
-            // Walk up the macro expansion chain till we reach a non-expanded span.
-            let mut scope_id = source_info.scope;
-            let mut span = source_info.span;
             let cm = self.fcx.ccx.sess().codemap();
-            while span.expn_id != NO_EXPANSION {
-                span = cm.source_callsite(span);
-            }
-            // Ditto for scope
-            while self.mir.visibility_scopes[scope_id].span.expn_id != NO_EXPANSION {
-                let parent_scope_id = self.mir.visibility_scopes[scope_id].parent_scope;
-                scope_id = match parent_scope_id {
-                    Some(id) => id,
-                    None => break // Can happen if the current function is defined inside a macro.
+            // Walk up the macro expansion chain until we reach a non-expanded span.
+            let mut span = source_info.span;
+            while span.expn_id != NO_EXPANSION && span.expn_id != COMMAND_LINE_EXPN {
+                if let Some(callsite_span) = cm.with_expn_info(span.expn_id,
+                                                    |ei| ei.map(|ei| ei.call_site.clone())) {
+                    span = callsite_span;
+                } else {
+                    break;
                 }
             }
-            let scope_metadata = self.scope_metadata_for_span(scope_id, span);
+            let scope_metadata = self.scope_metadata_for_loc(source_info.scope, span.lo);
```

Comment:
Can you add a comment that using the original scope here is not entirely correct, but benign, since it will always be a scope contained within the actual scope of the new span? Something to that effect. This is rather subtle, so I'd like to have a comment that draws some attention that something is going here.